### PR TITLE
chore: reorganize things

### PR DESCRIPTION
**Pull request by thoward27**
For Reviewers:

- Create a new clone of this repo (not your own fork and don't delete your current local clone)
- Switch to the dev branch
- Follow the contribution docs to install hermit and earthly.
- Try using hermit and earthly to install/compile the crates and run tests

### CODE REVIEW

**Comment by olwmc** [comment]
Lfg this looks good. I don't know how much I can comment on the actual changes to the repo (that'll get done after this) but I can give some notes for the instructions given in the PR:

* [X] Create a new clone of this repo (not your own fork and don't delete your current local clone)
    * `gh repo clone` 😎 
* [X] Switch to the dev 
* [X] Follow the contribution docs to install hermit and earthly.
    * Contrib docs should probably mention that earthly comes with hermit. I was a little confused looking for the link to install earthly. Also, the [shell hooks](https://docs.earthly.dev/guides/shell-hooks) link is broken. Do you mean [this link?](https://cashapp.github.io/hermit/usage/shell)
    * Hermit doesn't seem to be widely available through standard package managers (Except pip?). I have no clue how stable Hermit's development is but I immediately wonder if we'll have problems with divergent versions. I literally don't know a thing about devops though so I'm not going to pretend like I have any reason to believe this is something we should care about.
* [X] Try using hermit and earthly to install/compile the crates and run tests
    * If I just follow the instructions, I get the error below when trying to run `earthly +test`. This required A) starting the docker daemon (`sudo systemctl start docker`) and B) running earthly as `sudo` so it can access the docker socket. Not sure if this is some weird arch specific thing but I figured I should mention it. After that small fix things ran fine.
         * `_unknown *failed* | could not determine buildkit address - is Docker or Podman running?`
    * I'll note that, what I think is just compiling CLAM, was *significantly* slower than normal. I don't know if this is because we're doing it in a container or what but earthly took forever to compile CLAM (I have an i7, 32g RAM).
    * `========================== 🌍 Earthly Build  ✅ SUCCESS ==========================` !!

Great work. This looks good. Just some minor snags.

**Comment by thoward27** [comment]
@olwmc can you comment on re-running the compile step? I'd imagine the first run was building a lot of the cache components that your system may have already had in place. Subsequent compilations should be significantly faster. Please let me know if you don't see that performance shift

**Comment by olwmc** [comment]
@thoward27  Upon re-timing (and cargo cleaning) it's nearly identical (if I use the same command earthly uses). I think you're correct. Feel free to ignore my perf comment :)

File: CONTRIBUTING.md
Comment (morganprior) [inline CONTRIBUTING.md:15]: This didn't work for me without rebooting my whole computer so I think you should add a sentence telling them to reboot after this.

File: CONTRIBUTING.md
Comment (morganprior) [inline CONTRIBUTING.md:21]: I think the link for shell hooks here is wrong; I think you want https://cashapp.github.io/hermit/usage/shell/ instead 

Also, since I'm not really good with these things, it is nice to know why it's necessary for me to be doing them so maybe add a line like "Doing this will prevent you from needing to reactivate your hermit environment every time you re-open clam" (or something like that--not sure if my suggestion is totally accurate)

File: CONTRIBUTING.md
Comment (morganprior) [inline CONTRIBUTING.md:68]: Not urgent but someone should spell this out eventually.

Comment (morganprior) [reply CONTRIBUTING.md:68]: (I would if I remembered how to set up version bumping)

File: CONTRIBUTING.md
Comment (morganprior) [inline CONTRIBUTING.md:16]: This is mentioned in the link so you may not need to mention it here, but I did have to manually add to my path. I'm guessing most new additions to the research group will be more computer literate than I was when I joined, but I would've had _no idea_ how to add something to path when I joined the group. Also, even now, I didn't know if I should add '/home/username/bin' or '/home/username/bin/hermit' to my path. Maybe this will be obvious to other people or need not be dealt with here, just thought I'd mention it.

**Review by morganprior — CHANGES_REQUESTED** [review]
I was able to get this all set up, but there are a few changes I think you should make to CONTRIBUTING.md before giving this to new people in the group.

File: CONTRIBUTING.md
Comment (olwmc) [reply CONTRIBUTING.md:15]: I think you only need to re-login for this to work. Group memberships are checked at login. Rebooting will do this as well.

**Review by morganprior — APPROVED** [review]
LGTM!

**Review by ndaniels — COMMENTED** [review]
Right now, I'm not happy with the dependency on Docker. It's a complexity hellhole that breaks all the time. I know it's optional, but I'd like to see if we can avoid it. Is additional devops complexity like Earthly really needed? If it's just needed for Windows, fine, but given that it adds the dependency on make, which is itself hell... I'm not liking this particular approach.

"You will need this to test the CI/CD pipelines locally or to use Earthly to run various project commands.
You may need to start the docker daemon before running any commands. You can do this with sudo systemctl start docker.
You may need to add your user to the docker group to run docker commands without sudo. You can do this with sudo usermod -aG docker $USER. You need to re-login or reboot after doing this."

these instructions are *nix-specific, but I thought this Earthly/Docker stuff was needed to support Windows developers?

Someone with access to a Windows machine will need to test all this *in windows* before approval.

**Comment by djperrone** [comment]
I was able to get everything working without too much hassle but as one of the few windows users in the group I have to say I'm not sure why any of this is being advertised as helping Windows users. Hermit and Earthy do not work on Windows. WSL is required for any of this to work and getting Docker to run on WSL also requires additional steps. I also agree with Noah in that I am not a fan of having Docker or make required and am not entirely sure why earthly is necessary. All this being said, it was still easier to setup than our previous poetry virtual python environment system so there's that. Here is the link to the stackoverflow post for getting Docker to work on wsl2:  
https://askubuntu.com/questions/1379425/system-has-not-been-booted-with-systemd-as-init-system-pid-1-cant-operate

The top answer worked for me but is quite verbose. Essentially, I had to reinstall wsl (I already had wsl2 installed) and update it and then modify a config file to enable systemd. Here are the commands I ran:

- Powershell
    1. wsl --install
    2. wsl --update
    3. wsl --shutdown
    4. wsl

- wsl bash
    1. sudo -e /etc/wsl.conf
    2. *append the following two lines to the config file*
    [boot] 
    systemd=true

- Powershell
    1. wsl --shutdown
    2. wsl

- wsl bash
    1. sudo systemctl start docker

**Comment by nishaq503** [comment]
> I was able to get everything working without too much hassle but as one of the few windows users in the group I have to say I'm not sure why any of this is being advertised as helping Windows users. Hermit and Earthy do not work on Windows. WSL is required for any of this to work and getting Docker to run on WSL also requires additional steps. I also agree with Noah in that I am not a fan of having Docker or make required and am not entirely sure why earthly is necessary. All this being said, it was still easier to setup than our previous poetry virtual python environment system so there's that. Here is the link to the stackoverflow post for getting Docker to work on wsl2: https://askubuntu.com/questions/1379425/system-has-not-been-booted-with-systemd-as-init-system-pid-1-cant-operate
> 
> The top answer worked for me but is quite verbose. Essentially, I had to reinstall wsl (I already had wsl2 installed) and update it and then modify a config file to enable systemd. Here are the commands I ran:
> 
> * Powershell
>   
>   1. wsl --install
>   2. wsl --update
>   3. wsl --shutdown
>   4. wsl
> * wsl bash
>   
>   1. sudo -e /etc/wsl.conf
>   2. _append the following two lines to the config file_
>      [boot]
>      systemd=true
> * Powershell
>   
>   1. wsl --shutdown
>   2. wsl
> * wsl bash
>   
>   1. sudo systemctl start docker

The docs do state that hermit and earthly are opt in. I will make this more clear up front.

Can you open a PR to this dev branch to add the required WSL steps to the contributions guidelines?

**Comment by ndaniels** [comment]
I got everything to work without a hitch on my M1 Mini (the machine in my office). I won't install Docker on my laptop.

I would very much like to investigate https://github.com/cross-rs/cross in the future. Note: it still relies on Docker; there's no easy way around that for cross-compilation. But I'd much prefer to drive everything from Cargo, since this is primarily a Rust project (with ancillary Python bits). So, a Cargo plugin is probably going to be more reliable and less fragile in the long run.

I should note that for  developers, Docker/Earthly/Cross aren't necessary at all. You can rely on Github to do the cross-testing and such.  Indeed, if I allowed ITS at URI to have their way with my machine (which I never will) I wouldn't be allowed to use Docker locally (!!!)

**Comment by nishaq503** [comment]
> I got everything to work without a hitch on my M1 Mini (the machine in my office). I won't install Docker on my laptop.
> 
> 
> 
> I would very much like to investigate https://github.com/cross-rs/cross in the future. Note: it still relies on Docker; there's no easy way around that for cross-compilation. But I'd much prefer to drive everything from Cargo, since this is primarily a Rust project (with ancillary Python bits). So, a Cargo plugin is probably going to be more reliable and less fragile in the long run.
> 
> 
> 
> I should note that for  developers, Docker/Earthly/Cross aren't necessary at all. You can rely on Github to do the cross-testing and such.  Indeed, if I allowed ITS at URI to have their way with my machine (which I never will) I wouldn't be allowed to use Docker locally (!!!)

Excellent! Can you open an issue and add a card on the project board for looking at cross-rs? And, in the meantime, if you think that these docs are good enough, can you approve the PR?

**Review by ndaniels — APPROVED** [review]
Looks good, modulo my distaste about tools subsuming Cargo.

**Review by nishaq503 — APPROVED** [review]
LGTM!

File: CONTRIBUTING.md
Comment (carlstoker) [inline CONTRIBUTING.md:38]: I believe systemctl is part of systemd. It's part of most major Linux distros, but what about macOS?

File: CONTRIBUTING.md
Comment (carlstoker) [inline CONTRIBUTING.md:42]: Can we simplify this to `~/bin/` or `$HOME/bin/`?  

Easier instructions on adding to the path may be helpful as well. Something like
```
echo 'export PATH="$PATH:~/bin"' >> ~/.bashrc
source ~/.bashrc
```
or
```
echo 'export PATH="$PATH:~/bin"' >> ~/.zshrc
source ~/.zshrc
```

depending on the shell they are running.

File: CONTRIBUTING.md
Comment (carlstoker) [inline CONTRIBUTING.md:39]: Fun fact: I did this and then rebooted with `sudo reboot now` from my shell. When I rebooted, the new group did not activate. `newgrp` worked, but it changes your primary group (probably not desired for most people).

**Review by carlstoker — COMMENTED** [review]
Assuming nothing, I tested the new contributing steps with a fresh install of Debian 12.

I ran into a few issues with the steps for contributing.

I don't know how much hand-holding we want to do. Some comments are more pedantic than others.

**Comment by nishaq503** [comment]
> Assuming nothing, I tested the new contributing steps with a fresh install of Debian 12.
> 
> I ran into a few issues with the steps for contributing.
> 
> I don't know how much hand-holding we want to do. Some comments are more pedantic than others.

I have incorporated most of your comments and suggested changes.

I consider Debian as a more advanced distro than Ubuntu, which does have certain things pre-installed. As for hand-holding, the intended audience is the typical URI CSC440 student being added to the research group; someone with little to no experience developing a large project with multiple people, and who needs to be onboarded quickly so they can start working on their research project for a semester/year/arbitrarily-long-time-span.

**Pull request merged**
